proxy/modules/native_module_host/src/native_module_host.c

```diff
@@ -0,0 +1,418 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+#include "azure_c_shared_utility/gballoc.h"
+#include "azure_c_shared_utility/xlogging.h"
+#include "azure_c_shared_utility/macro_utils.h"
+
+// replace one or more with remote_module.h??
+#include "gateway.h"
+#include "broker.h"
+#include "module_loader.h"
+#include "message.h"
+#include "experimental/event_system.h"
+#include "module.h"
+#include "module_access.h"
+#include "module_loaders/dynamic_loader.h"
+#include "native_module_host.h"
+#include "parson.h"
+
+
+#define LOADER_NAME_KEY "name"
+#define LOADER_ENTRYPOINT_KEY "entrypoint"
+
+
+typedef struct MODULE_HOST_TAG
+{
+	MODULE_LIBRARY_HANDLE module_library_handle;
+	const MODULE_LOADER* module_loader;
+	MODULE_HANDLE module;
+	BROKER_HANDLE module_host_broker;
+} MODULE_HOST;
+
+// Arguments coming to this module are the module_args coming from the _Create_ control message.
+// in this circumstance, all we are goint to do is copy the string to send to our Module_Create function.
+static void* NativeModuleHost_ParseConfigurationFromJson(const char* configuration)
+{
+	char* config_str;
+	if (configuration == NULL)
+	{
+		/*Codes_SRS_NATIVEMODULEHOST_17_003: [ NativeModuleHost_ParseConfigurationFromJson shall return NULL if configuration is NULL. ]*/
+		config_str = NULL;
+	}
+	else
+	{
+		/*Codes_SRS_NATIVEMODULEHOST_17_004: [ On success, NativeModuleHost_ParseConfigurationFromJson shall return a valid copy of the configuration string given. ]*/
+		if (mallocAndStrcpy_s(&config_str, configuration) != 0)
+		{
+			/*Codes_SRS_NATIVEMODULEHOST_17_005: [ NativeModuleHost_ParseConfigurationFromJson shall return NULL if any system call fails. ]*/
+			config_str = NULL;
+		}
+	}
+    return config_str;
+}
+
+// Free the string we copied.
+static void NativeModuleHost_FreeConfiguration(void* configuration)
+{
+	/*Codes_SRS_NATIVEMODULEHOST_17_006: [ NativeModuleHost_FreeConfiguration shall do nothing if passed a NULL configuration. ]*/
+    if (configuration != NULL)
+    {
+		/*Codes_SRS_NATIVEMODULEHOST_17_007: [ NativeModuleHost_FreeConfiguration shall free configuration. ]*/
+        free((char*)configuration);
+    }
+}
+
+static int parse_loader(JSON_Object* loader_json, GATEWAY_MODULE_LOADER_INFO* loader_info)
+{
+    int result;
+
+    // get loader name; we assume it is "native" if the JSON doesn't have a name
+	/*Codes_SRS_NATIVEMODULEHOST_17_013: [ NativeModuleHost_Create shall get "name" string from the "outprocess.loader" object. ]*/
+    const char* loader_name = json_object_get_string(loader_json, LOADER_NAME_KEY);
+    if (loader_name == NULL)
+    {
+		/*Codes_SRS_NATIVEMODULEHOST_17_014: [ If the loader name string is not present, then NativeModuleHost_Create shall assume the loader is the native loader name. ]*/
```

Comment:
>native loader name. [](start = 143, length = 19)

Shall we change this requirement from native loader name to Dynamic loader name?